server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java

```diff
@@ -209,39 +204,60 @@ public DiskThresholdSettings getDiskThresholdSettings() {
         }
     }
 
-    private record CachedAllocationStats(Map<String, NodeAllocationStats> stats, long timestampMillis) {}
-
-    private static class AllocationStatsCache {
+    private static class AllocationStatsCache extends CancellableSingleObjectCache<Long, Long, Map<String, NodeAllocationStats>> {
         private volatile long ttlMillis;
         private final ThreadPool threadPool;
-        private final AtomicReference<CachedAllocationStats> cachedStats;
+        private final ExecutorService executorService;
```

Comment:
nit: could this be an `Executor`? No need for it to be `AutoCloseable` (it's a disaster if you try and close one of the threadpool's executors)